tensorflow_datasets/core/utils/dummy_beam.py

```diff
@@ -0,0 +1,44 @@
+# coding=utf-8
+# Copyright 2020 The TensorFlow Datasets Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Utilities for beam wrapper."""
+
+import types
+
+_MSG = ("Failed importing {name}. Please install {name}."
+        " Using pip install {name}")
+
+class Empty(): # pylint: disable=too-few-public-methods
+    """Empty class for beam API."""
+    def __call__(self, *args, **kwargs):
+        return None
+
+class DummyBeam(types.ModuleType): # pylint: disable=too-few-public-methods
+    """Dummy class.
+    Raise:
+      ImportError, when calling beam API and apache_beam was not installed.
+    """
+    DoFn = Empty
+    Pipeline = Empty
+
+    def __init__(self):
+        self.module_name = "apache_beam"
+
+    def __getattribute__(self, _):
+        if getattr(DummyBeam, _, None) is Empty:
+            _name = super().__getattribute__('module_name')
+            err_msg = _MSG.format(name=_name)
+            raise ImportError(err_msg)
```

Comment:
Define `__getattr__` instead of `__getattribute__`.

Reply:
`__getattr__` make some trouble here because we want `DoFn ` to call with DummyBeam so if use `getattr `here this will executing without any errors here, the reason why I am using `__getattribute__` is `__getattribute__` calls no matter what unlike `getattr `

Reply:
I don't think we should throw an error when `DoFn` is accessed.

> If beam isn't found, then the imports works, but calling SomeFn(), some_ptransform_fn(),... raise the lazy import error ('Please install appache_beam')

We don't want this to happen. So, we should not throw any errors when members of `DummyBeam` are accessed.

However, proceed only after consulting @Conchylicultor. 
(Just in case I am wrong :sweat_smile:)

Reply:
Beam API have DoFn and pipiline but it dont have SomeFn its just name i think thats why it inherit beam.DoFn so i think it is good to get error when invoking beam.DoFn 🙂

Reply:
@vijayphoenix is right. The all point of this module is to allow using beam without raising error when declaring the beam functions. However error should be raised when the functions are called.

Reply:
I think Using getattr instead getattribute is also not worked as getattr calls only when it don't fine DoFn or Pipeline which is called insider class DummyBeam, can you suggest some other way ?

Reply:
Add some like 
```
class Empty():
    def __getattr__(self, _)
        raise Error
```

Reply:
Thanks but this won't works 🙂

Reply:
However I found a way to do this and its working when  I call SomeFn it raises error, is it necessary to use `types.ModuleType` ?

Reply:
You shouldn't have to use `__new__`. `__init__` should be enough.

What is the issue with `types.ModuleType` ?

Reply:
Modules should inherit from `ModuleType`, but not other Class or functions.


```
class EmptyClass(object):
  def __init__(self):
    raise ...
```

Reply:
okay then doing in this way ?
```py
class Empty(object):
   
    def __init__(self, *args, **kwargs):
        err_msg = _MSG.format(name="apache_beam")
        raise ImportError(err_msg)
        
    
class DummyBeam(types.ModuleType):

    DoFn = Empty
    Pipeline = Empty
    ptransform = Empty

    def __init__(self):
        return None
```
what is use of `types.ModuleType` here ?